contracts/yellow/YellowClearingBase.sol

```diff
@@ -0,0 +1,207 @@
+
+    // Next implementation
+    YellowClearingBase private _nextImplementation;
+
+    // Address of this contract
+    address private immutable _self = address(this);
+
+    // Constructor
+    constructor(YellowClearingBase previousImplementation) {
+        _grantRole(DEFAULT_ADMIN_ROLE, msg.sender);
+        _grantRole(REGISTRY_MAINTAINER_ROLE, msg.sender);
+
+        // Review: introduce a helper `isValidAddress(address)` method
+        if (address(previousImplementation) != address(0)) {
+            _grantRole(PREVIOUS_IMPLEMENTATION_ROLE, address(previousImplementation));
+        }
+    }
+
+    // Get next implementation address
+    function getNextImplementation() external view returns (YellowClearingBase) {
+        return _nextImplementation;
+    }
+
+    // Set next implementation address
+    function setNextImplementation(YellowClearingBase nextImplementation)
+        external
+        onlyRole(REGISTRY_MAINTAINER_ROLE)
+    {
+        require(address(_nextImplementation) == address(0), 'Next implementation already set');
+        require(
+            address(nextImplementation) != address(0) && address(nextImplementation) != _self,
+            'Invalid nextImplementation supplied'
+        );
+
+        require(
+            nextImplementation.hasRole(PREVIOUS_IMPLEMENTATION_ROLE, address(this)),
+            'Previous implementation role is absent'
+        );
+
+        _nextImplementation = nextImplementation;
+
+        emit NextImplementationSet(nextImplementation);
+    }
+
+    // Has participant
+    function hasParticipant(address participant) public view returns (bool) {
+        return _participantData[participant].status != ParticipantStatus.None;
+    }
+
+    // Get participant data
+    function getParticipantData(address participant)
+        external
+        view
+        returns (ParticipantData memory)
+    {
+        require(hasParticipant(participant), 'Participant does not exist');
+
+        return _participantData[participant];
+    }
+
+    // Recursively check that participant is not present in the registry
+    function requireParticipantNotRegistered(address participant) public view {
+        if (address(_nextImplementation) != address(0)) {
+            _nextImplementation.requireParticipantNotRegistered(participant);
+        }
+
+        require(!hasParticipant(participant), 'Participant already registered');
+    }
+
+    // Register participant with Pending status using signature by its Broker
+    function registerParticipant(address participant, ParticipantRegisterData calldata rData)
+        external
+    {
+        requireParticipantNotRegistered(participant);
+
+        require(
+            keccak256(abi.encode(participant)).toEthSignedMessageHash().recover(
+                rData.vaultBrokerSignature
+            ) == rData.vault.getBrokerAddress(),
+            'Signer is not participant vault broker'
+        );
```

Comment:
Only require this if Vault address is specified.
Current workflow here is to allow participant that do not have a Vault, to still be able to register (like trading bots)

Reply:
This would allow anyone to flood the contract supplying random addresses as `participant` and `address(0)` as `vault`, which contradicts a purpose we supply vault broker signature to this method for.

Right now I don't have a solution to this. Maybe you do?